src/openrct2/ReplayManager.cpp

```diff
@@ -0,0 +1,775 @@
+        GameAction::Ptr action;
+        uint8_t playerid = 0;
+        uint8_t callback = 0;
+        uint32_t commandIndex = 0;
+
+        bool operator<(const ReplayCommand& comp) const
+        {
+            // First sort by tick
+            if (tick < comp.tick)
+                return true;
+            if (tick > comp.tick)
+                return false;
+
+            // If the ticks are equal sort by commandIndex
+            return commandIndex < comp.commandIndex;
+        }
+    };
+
+    struct ReplayRecordData
+    {
+        MemoryStream parkData;
+        MemoryStream spriteSpatialData;
+        MemoryStream parkParams;
+        std::string name;      // Name of play
+        uint64_t timeRecorded; // Posix Time.
+        uint32_t tickStart;    // First tick of replay.
+        uint32_t tickEnd;      // Last tick of replay.
+        std::multiset<ReplayCommand> commands;
+        std::vector<std::pair<uint32_t, rct_sprite_checksum>> checksums;
+        uint32_t checksumIndex;
+    };
+
+    class ReplayManager final : public IReplayManager
+    {
+        enum class ReplayMode
+        {
+            NONE = 0,
+            RECORDING,
+            PLAYING,
+            NORMALISATION,
+        };
+
+    public:
+        virtual ~ReplayManager()
+        {
+        }
+
+        virtual bool IsReplaying() const override
+        {
+            return _mode == ReplayMode::PLAYING;
+        }
+
+        virtual bool IsRecording() const override
+        {
+            return _mode == ReplayMode::RECORDING;
+        }
+
+        virtual bool IsNormalising() const override
+        {
+            return _mode == ReplayMode::NORMALISATION;
+        }
+
+        virtual void AddGameCommand(
+            uint32_t tick, uint32_t eax, uint32_t ebx, uint32_t ecx, uint32_t edx, uint32_t esi, uint32_t edi, uint32_t ebp,
+            uint8_t callback) override
+        {
+            if (_currentRecording == nullptr)
+                return;
+
+            uint32_t args[7];
+            args[0] = eax;
+            args[1] = ebx;
+            args[2] = ecx;
+            args[3] = edx;
+            args[4] = esi;
+            args[5] = edi;
+            args[6] = ebp;
+
+            _currentRecording->commands.emplace(gCurrentTicks, args, callback, _commandId++);
+
+            // Force a checksum record the next tick.
+            _nextChecksumTick = tick + 1;
+        }
+
+        virtual void AddGameAction(uint32_t tick, const GameAction* action) override
+        {
+            if (_currentRecording == nullptr)
+                return;
+
+            MemoryStream stream;
+            DataSerialiser dsOut(true, stream);
+            action->Serialise(dsOut);
+
+            std::unique_ptr<GameAction> ga = GameActions::Create(action->GetType());
+            ga->SetCallback(action->GetCallback());
+
+            stream.SetPosition(0);
+            DataSerialiser dsIn(false, stream);
+            ga->Serialise(dsIn);
```

Comment:
I don't understand the purpose of the code here. You create a stack variable `stream` and you serialise the currently invoked action from param into it, then create a copy of this game action, but only with callback set, which you then use to overwrite the previously serialised… OH, it's not overwriting, but _reading_ from the serialised stream? So you basically copy the game action? Have I understood correctly?

I had such a hard time wrapping my head around this and needed to cross-reference so many things, there should be a better way to express it in code or at least provide some comments explaining what it does.

Reply:
e.g. this could be named `Deserialise` or use a static method to `CreateFromSerialisedStream`

Reply:
If I touch the DataSerialiser code now I will affect a lot of game action, but yes I can clean up that part so it makes more sense.